src/app-layer-dhcp.c

```diff
@@ -0,0 +1,699 @@
+/* Copyright (C) 2015 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file DHCP application layer detector and parser.
+ *
+ * \author Tom DeCanio <decanio.tom@gmail.com>
+ */
+
+#include "suricata-common.h"
+#include "stream.h"
+#include "conf.h"
+
+#include "util-print.h"
+#include "util-unittest.h"
+
+#include "app-layer-detect-proto.h"
+#include "app-layer-parser.h"
+
+#include "app-layer-dhcp.h"
+
+//#define PRINT
+
+/* The default port to probe for echo traffic if not provided in the
+ * configuration file. */
+#define DHCP_DEFAULT_SERVER_PORT "67"
+#define DHCP_DEFAULT_CLIENT_PORT "68"
+
+#define DHCP_MIN_FRAME_LEN 232
+
+/* Enum of app-layer events for an echo protocol. Normally you might
+ * have events for errors in parsing data, like unexpected data being
+ * received. For echo we'll make something up, and log an app-layer
+ * level alert if an empty message is received.
+ *
+ * Example rule:
+ *
+ * alert dhcp any any -> any any (msg:"SURCATA dhcp empty message"; \
+ *    app-layer-event:dhcp.empty_message; sid:X; rev:Y;)
+ */
+enum {
+    DHCP_DECODER_EVENT_EMPTY_MESSAGE,
+};
+
+SCEnumCharMap dhcp_decoder_event_table[] = {
+    {"EMPTY_MESSAGE", DHCP_DECODER_EVENT_EMPTY_MESSAGE},
+};
+
+static uint32_t dhcp_config_max_transactions = 32;
+
+static DHCPGlobalState dhcpGlobalState;
+
+static void DHCPTxFree(DHCPGlobalState *dhcp, void *tx, uint32_t locked)
```

Comment:
not a big fan of the conditional locking, esp since it's called with 0/1 so at the callers it's unclear what it means. Perhaps a unlocked func and lock wrapper would make sense, although it would have to be done such that all the actual freeing isn't blocking other threads.